ui/tickets_activity_page.go

```diff
@@ -0,0 +1,162 @@
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/planetdecred/godcr/wallet"
+
+	"gioui.org/layout"
+	"gioui.org/text"
+	"github.com/planetdecred/godcr/ui/decredmaterial"
+	"github.com/planetdecred/godcr/ui/values"
+)
+
+const PageTicketsActivity = "TicketsActivity"
+
+type ticketsActivityPage struct {
+	th           *decredmaterial.Theme
+	tickets      **wallet.Tickets
+	ticketsList  layout.List
+	filterSorter int
+
+	orderDropDown      *decredmaterial.DropDown
+	ticketTypeDropDown *decredmaterial.DropDown
+	walletDropDown     *decredmaterial.DropDown
+	common             pageCommon
+}
+
+func (win *Window) TicketActivityPage(c pageCommon) Page {
+	pg := &ticketsActivityPage{
+		th:          c.theme,
+		common:      c,
+		tickets:     &win.walletTickets,
+		ticketsList: layout.List{Axis: layout.Vertical},
+	}
+	pg.orderDropDown = c.theme.DropDown([]decredmaterial.DropDownItem{{Text: "Newest"}, {Text: "Oldest"}}, 1)
+	pg.ticketTypeDropDown = c.theme.DropDown([]decredmaterial.DropDownItem{
+		{Text: "All"},
+		{Text: "Unmined"},
+		{Text: "Immature"},
+		{Text: "Live"},
+		{Text: "Voted"},
+		{Text: "Missed"},
+		{Text: "Expired"},
+		{Text: "Revoked"},
+	}, 1)
+
+	return pg
+}
+
+func (pg *ticketsActivityPage) Layout(gtx layout.Context) layout.Dimensions {
+	c := pg.common
+	body := func(gtx C) D {
+		page := SubPage{
+			title: "Ticket activity",
+			back: func() {
+				c.changePage(PageTickets)
+			},
+			body: func(gtx C) D {
+				walletID := c.info.Wallets[pg.walletDropDown.SelectedIndex()].ID
+				tickets := (*pg.tickets).Confirmed[walletID]
+				return layout.Stack{Alignment: layout.N}.Layout(gtx,
+					layout.Expanded(func(gtx C) D {
+						return layout.Inset{Top: values.MarginPadding60}.Layout(gtx, func(gtx C) D {
+							return c.theme.Card().Layout(gtx, func(gtx C) D {
+								gtx.Constraints.Min = gtx.Constraints.Max
+								if pg.ticketTypeDropDown.SelectedIndex()-1 != -1 {
+									tickets = filterTickets(tickets, func(ticketStatus string) bool {
+										return ticketStatus == strings.ToUpper(pg.ticketTypeDropDown.Selected())
+									})
+								}
+
+								if len(tickets) == 0 {
+									txt := c.theme.Body1("No tickets yet")
+									txt.Color = c.theme.Color.Gray2
+									txt.Alignment = text.Middle
+									return layout.Inset{Top: values.MarginPadding15}.Layout(gtx, func(gtx C) D { return txt.Layout(gtx) })
+								}
+								return layout.UniformInset(values.MarginPadding16).Layout(gtx, func(gtx C) D {
+									return pg.ticketsList.Layout(gtx, len(tickets), func(gtx C, index int) D {
+										return ticketRowActivity(gtx, &c, tickets[index], index)
+									})
+								})
+							})
+						})
+					}),
+					layout.Stacked(func(gtx C) D {
+						return pg.dropDowns(gtx)
+					}),
+				)
+			},
+		}
+		return c.SubPageLayout(gtx, page)
+	}
+
+	return c.Layout(gtx, func(gtx C) D {
+		return c.UniformPadding(gtx, body)
+	})
+}
+
+func (pg *ticketsActivityPage) dropDowns(gtx layout.Context) layout.Dimensions {
```

Comment:
The wallet dropdown, sort dropdown and a page specific dropdown seems to be used in several pages. They could be implemented as "components" that can be called from each of those pages. This eliminates repetition.